### ISSUE TITLE
Fatal Error During SuperCollider Server Boot

### ISSUE THREAD

**Issue by gbochmann** [overtone/overtone#394]
Hi!

I tried running overtone with an internal server using:
`(use 'overtone.live)`

This results in the following fatal error:
```
--> Loading Overtone...
  **********************************************************
       WARNING: JVM argument TieredStopAtLevel=1 is active, and may
       lead to reduced performance. This happens to currently be the
       default lein setting:

       https://github.com/technomancy/leiningen/pull/1230

       If you didn't intend this JVM arg to be specified, you can turn
       it off in your project.clj file or your global
       ~/.lein/profiles.clj file by adding the key-val

       :jvm-opts ^:replace []
       **********************************************************
--> Booting internal SuperCollider server...
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000012f96857c, pid=8832, tid=0x0000000000014307
#
# JRE version: Java(TM) SE Runtime Environment (8.0_121-b13) (build 1.8.0_121-b13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.121-b13 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libscsynth.dylib+0x557c]  _ZN12SC_UdpInPortC2EP5Worldi+0x1be
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/gbochmann/dev/overtone/tutorial/hs_err_pid8832.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Found 0 LADSPA plugins
*** ERROR: open directory failed '/Users/gbochmann/Library/Application Support/SuperCollider/synthdefs'
Number of Devices: 3
   0 : "Built-in Microph"
   1 : "Built-in Output"
   2 : "HDMI"

input and output sample rates do not match. 44100 != 48000
could not initialize audio.
SocketException The transport's socket appears to have lost its connection to the nREPL server
        clojure.tools.nrepl.transport/bencode/fn--10199/fn--10200 (transport.clj:95)
        clojure.tools.nrepl.transport/bencode/fn--10199 (transport.clj:95)
        clojure.tools.nrepl.transport/fn-transport/fn--10171 (transport.clj:42)
Exception in thread "Thread-3"  clojure.core/binding-conveyor-fn/fn--4676 (core.clj:1938)
clojure.lang.ExceptionInfo: Subprocess failed {:exit-code 134}
        at clojure.core$ex_info.invokeStatic(core.clj:4617)
        at clojure.core$ex_info.invoke(core.clj:4617)
        at leiningen.core.eval$fn__5732.invokeStatic(eval.clj:264)
        at leiningen.core.eval$fn__5732.invoke(eval.clj:260)
        at clojure.lang.MultiFn.invoke(MultiFn.java:233)
        at leiningen.core.eval$eval_in_project.invokeStatic(eval.clj:366)
        at leiningen.core.eval$eval_in_project.invoke(eval.clj:356)
        at leiningen.repl$server$fn__11838.invoke(repl.clj:243)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:646)
        at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1881)
        at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1881)
        at clojure.lang.RestFn.invoke(RestFn.java:425)
        at clojure.lang.AFn.applyToHelper(AFn.java:156)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.core$apply.invokeStatic(core.clj:650)
        at clojure.core$bound_fn_STAR_$fn__4671.doInvoke(core.clj:1911)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.lang.Thread.run(Thread.java:745)
        java.util.concurrent.FutureTask.run (FutureTask.java:266)
        java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1142)
        java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:617)
        java.lang.Thread.run (Thread.java:745)
Bye for now!
```

Let me know, if you need any more info.

**Comment by hlolli** [comment]
Think this is the problem
```
input and output sample rates do not match. 44100 != 48000
could not initialize audio.
```
Your jack config is probably set to 44100 while supercollider defaults to 48000. You could either change your Jack settings or add SC startup file: http://doc.sccode.org/Reference/StartupFile.html

Easier still, let Overtone configure Supercollider for you by creating this file with the settings you wish to modify https://github.com/overtone/overtone/blob/master/docs/config.clj

**Comment by minuskruste** [comment]
I'm not using Jack as far as I can see. I'm on a mac running macOS 10.12.16. However, using an SC startup file still seems like a useful solution. I won't get to it today. Probably tomorrow. I'll let you know what happened.

**Comment by hlolli** [comment]
One thing I may want to add, if you are booting internal server, the SC startup file may get ignored, but I'm not really sure. It's actually safer to implement overtone config, maybe you have autogenerated ~/.overtone dir on your computer? (note the dot in the filename which causes it to be hidden in the Finder).

**Comment by Chovin** [comment]
I can't for the life of me figure out how to get the `~/.overtone/config.clj` to change the sample rate.
It sounded like you were saying we should have overtone change supercollider's sample rate to 44100. I tried both 44100 and 48000 to no avail.

my config.clj:
```clj
{:os :mac,
 :user-name "Chovin",
 :server :internal,
 :sc-args {:hw-sample-rate 44100},
 :versions-seen #{"v0.10.2"}}

```

any help would be appreciated

**Comment by hlolli** [comment]
Looking at the source code it seems as if the key `:mPreferredSampleRate` will set the sample rate for the internal sc-synth. That's a mouthful for a common operation. Can you give it a try? Alternatively it's always possible to configure jack with ~/.jackrc file as described https://github.com/overtone/overtone/wiki/Installing-and-starting-jack

**Comment by hlolli** [comment]
I tried it and it didn't work on my computer at least... let me look at this better.

**Comment by hlolli** [comment]
What's going on here in Overtone is that before the internal server is started, which happens only after jack has successfully started, it finds the sample rate of the jack instance and boots supercollider to that sample rate, therefore telling Overtone about sample rate is nonesense on my behalf.

About the problem pointed out from @minuskruste it's related to this problem here https://github.com/samaaron/sonic-pi/issues/87 since Sonic-pi uses the same SC-API (the same developers of Sonic-Pi and Overtone) it seems there has to be made a speical case for machintosh and CoreAudio.

**Comment by Chovin** [comment]
not quite sure how to change the sample rate of CoreAudio, but adding a http://doc.sccode.org/Reference/StartupFile.html does fix it, although I would prefer not setting that for SC globally

Is Jack installed automatically on mac and windows? (I'm on a mac)

**Comment by kwabenareuben6** [comment]
hmmm.

**Comment by hlolli** [comment]
@Chovin you could try setting that sample rate inside your ~/.overtone/config.clj under the key :sc-args like this
```clojure
{...
:sc-args {:hw-sample-rate 44100}
... }
```
@gbochmann I just did major changes to the internal server, can you test it if my changes fix your issue, what you'd need to do to test it, would be to pull the latest master, make sure that there are no `target` and `native` directories within your overtone directory before you start it (to prevent conflicts to older scsynth binaries) via lein repl (or cider if you're on emacs). And report if you encounter error or not?

**Comment by NightMachinary** [comment]
@hlolli From the latest master just downloaded:
```
--> Loading Overtone...
--> Booting internal SuperCollider server...
Found 0 LADSPA plugins
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000011e6546f0, pid=8186, tid=0x000000000001652f
#
# JRE version: OpenJDK Runtime Environment (8.0_152-b16) (build 1.8.0_152-b16)
# Java VM: OpenJDK 64-Bit Server VM (25.152-b16 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libscsynth.dylib+0x1f6f0]  _ZN7scsynth12SC_UdpInPortC2EP5WorldNSt3__112basic_stringIcNS3_11char_traitsIcEENS3_9allocatorIcEEEEi+0x1d0
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Base/- Code/Misc/overtone/hs_err_pid8186.log
#
# If you would like to submit a bug report, please visit:
#   http://www.azulsystems.com/support/
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Number of Devices: 3
   0 : "Built-in Microph"
   1 : "Built-in Output"
   2 : "BoomAudio"

ERROR: Input sample rate is 48000, but output is 44100. Mismatched sample rates are not supported. To disable input, set the number of input channels to 0.
could not initialize audio.
SocketException The transport's socket appears to have lost its connection to the nREPL server
Exception in thread "Thread-1" clojure.lang.ExceptionInfo: Subprocess failed {:exit-code 134}
	at clojure.core$ex_info.invokeStatic(core.clj:4617)
	at clojure.core$ex_info.invoke(core.clj:4617)
	at leiningen.core.eval$fn__4134.invokeStatic(eval.clj:264)
	at leiningen.core.eval$fn__4134.invoke(eval.clj:260)
	at clojure.lang.MultiFn.invoke(MultiFn.java:233)
	at leiningen.core.eval$eval_in_project.invokeStatic(eval.clj:366)
	at leiningen.core.eval$eval_in_project.invoke(eval.clj:356)
	at leiningen.repl$server$fn__5864.invoke(repl.clj:244)
	at clojure.lang.AFn.applyToHelper(AFn.java:152)
	at clojure.lang.AFn.applyTo(AFn.java:144)
	at clojure.core$apply.invokeStatic(core.clj:646)
	at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1881)
	at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1881)
	at clojure.lang.RestFn.invoke(RestFn.java:425)
	at clojure.lang.AFn.applyToHelper(AFn.java:156)
	at clojure.lang.RestFn.applyTo(RestFn.java:132)
	at clojure.core$apply.invokeStatic(core.clj:650)
	at clojure.core$bound_fn_STAR_$fn__4671.doInvoke(core.clj:1911)
	at clojure.lang.RestFn.invoke(RestFn.java:397)
	at clojure.lang.AFn.run(AFn.java:22)
	at java.lang.Thread.run(Thread.java:748)
	clojure.tools.nrepl.transport/bencode/fn--5154/fn--5155 (transport.clj:95)
	clojure.tools.nrepl.transport/bencode/fn--5154 (transport.clj:95)
	clojure.tools.nrepl.transport/fn-transport/fn--5126 (transport.clj:42)
	clojure.core/binding-conveyor-fn/fn--4676 (core.clj:1938)
	java.util.concurrent.FutureTask.run (FutureTask.java:266)
	java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
	java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
	java.lang.Thread.run (Thread.java:748)
Bye for now!
```

**Comment by hlolli** [comment]
@NightMachinary 

```
ERROR: Input sample rate is 48000, but output is 44100. Mismatched sample rates are not supported. To disable input, set the number of input channels to 0.
could not initialize audio.
```
In your mac go to  Applications > Utilities and change the sample rates to match, both for microphones and speakers.

**Comment by oubiwann** [comment]
For me, I had to change the settings in my DAW for the song I had created (since the DAW communicates with the Apollo Twin which is the sound device I'm using for Overtone).

**Comment by pastafari** [comment]
I got both Sonic Pi and Overtone to work with a Bluetooth speaker on OS X Sierra, and wanted to share my experience! 

1. What @hlolli suggested (both the Audio MIDI setup and the config for supercollider) works well if you're using the built-in speakers. 
2. I was still having trouble with an external Bluetooth speaker. Note that the error log includes this message:  ```To disable input, set the number of input channels to 0.```

So I went to Sound preferences and changed my input to internal microphone and output to external Bluetooth speaker, and voila everything worked.

Can't claim to say I know what I'm doing, just wanted to leave this here as a note to someone who lands up from Google (and maybe to future self!)

**Comment by brethorne** [comment]
Was running into a similar issue as above with
`ERROR: Input sample rate is 16000, but output is 48000. Mismatched sample rates are not supported. To disable input, set the number of input channels to 0.` 
Setting my speakers and mic to both 44100 makes it work. Thanks for the tips!

**Comment by plexus** [comment]
Overtone no longer contains an internal supercollider, we always start an external one (see https://groups.google.com/g/overtone/c/qndjDV5FS9Y ). So as long as you can install and run scsynth, things should work with Overtone as well.

So I think we can close this, please open a new issue if you still encounter problems.
